### ISSUE TITLE
NAT test failed dispite port-forwarded

### ISSUE THREAD

**Issue by SodaWithoutSparkles** [BiglySoftware/BiglyBT#2933]
_related: [reddit post 1](https://www.reddit.com/r/torrents/comments/12g9x05/biglybt_local_connection_never_connects/), [reddit post 2](https://www.reddit.com/r/BiglyBT/comments/12gfb14/biglybt_local_connection_never_connects/)_

## Setup
- Rabpberry pi 4
- `Linux raspberrypi 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64`
- BiglyBT stable 3.3.0.0 and beta 3.3.0.1_B27
- Router: TP-link TL-WDR7660 Gigabit
- Connection: ethernet
- no IPv6 connectivity
- Pi-hole is the DNS resolver, with cloudflare upstream

## Expected behaviour
- BiglyBT seeds normally
- TCP inbound connection success

## Actual behaviour
- BiglyBT seeds normally to connections considered "remote"
- all connections considered "local" (you established the connection) stays in State: Connecting with 0B/s up and down, grey Pieces bar
- UDP inbound success
- Network status test shows the following: 
```
<publicIP>:60000 (your computer) refused
    Check your port forwarding for TCP 60000
```
(I have redacted my publicIP for obvious reasons)

## What I have tried
1. Changing port to something else (54321/TCP)
    - No difference in result, except port changed
    - revert to port 60000/TCP

2. Manually forward port myself instead of using upnp
    - biglyBT now complains about unable to setup port forwarding because I have done so manually
    - no other changes
    - undo the manual port forwarding

3. Update to latest beta version
    - nothing changes
    - stays at beta

4. Use different methods to test if port is opened
    - canyouseeme.org
        - test for port 60000 shows opened
        - test for port 60001 shows not opened (as control)
    - nmap
        - TCP SYN ping `-PS` returns host is up
        - TCP ACK ping `-PA` returns host seems down
        - force TCP connect scan `-sT` returns 60000/TCP opened
        - force UDP scan `-sU` returns `60000/udp open|filtered`
        - force TCP SYN scan `-sS` returns 60000/TCP opened

5. Use telnet to see if port is opened
    - both mobile data and home WiFi network shows the same results

```
$ telnet <publicIP> 60000
Connected to <publicIP>.
Escape character is '^]'.
```

6. Disabled pi-hole blocking
    - nothing was blocked anyways
    - no differnece in behaviour
    - enabled again

7. Switched pi-hole upstream to google
   - as I saw NOTIMP (not implemented) response to query `nettest.biglybt.com` type `ANY`
    - cloudflare seems to block type `ANY` to prevent DNS amplification attacks
    - no difference in behaviour
    - switched back to cloudflare

8. Switched I2P implementation to biglyBT internal rather than external i2pd router with i2cp
    - no differnece in behaviour
    - switched back to i2pd external

9. Enabled all networks
    - networks enabled: 
        - Public IP network
        - I2P network
        - Tor network
    - no difference in behaviour
    - stays at all networks enabled

10. NAT / Server port test 
    - Clicked "Apply" after testing 60000/TCP
    - no difference in behaviour
```
Testing port TCP 60000 ... 
OK ! (<publicIP>)
Congratulations, you fixed it!
	IPv6 not available
```

## Plugins that may be related
- I2P helper
- tor helper
- location provider
- mlDHT
- Swarm discoveries
- upnp media server
- utp plugin
- webtorrent support

## Torrents seeding
- `ubuntu-22.10-desktop-amd64.iso`
    - from ubuntu's official site
- `linuxmint-21.1-cinnamon-64bit.iso`
    - from linux mint's official site
- `archlinux-2023.04.01-x86_64.iso`
    - from arch linux's official site
- `core_BiglyBT3301-B27-signed.jar`
    - from updating from stable to beta version

## screenshots
![Screenshot](https://user-images.githubusercontent.com/64138578/230822648-2933251e-0a1c-44e1-a10f-b5077ccce095.jpg)

**Comment by SodaWithoutSparkles** [comment]
I also saw a warning about "illegal reflective access" when launching the stable version off from the bash command line. Would it be related?

**Comment by parg** [comment]
No, ignore that

**Comment by parg** [comment]
Well the NAT test works perfectly well for me and many others so I would assume that the issue is something to do with your setup. 

You wouldn't believe the number of people who turn up with NAT issues, I can't spend resources investigating individual cases as I am the only active developer on BiglyBT and I also have a life to try and lead.

You sound like you know what you're doing - please keep trying to figure it out!

**Comment by SodaWithoutSparkles** [comment]
Is there any general directions? I think I am out of ideas. The interesting thing is that dispite telnet and canyouseeme.org both said port opened, and the NAT/firewall test shows opened too, the network status still said refused. 

Can you tell me how biglyBT test for the port being opened or not? I remember seeing a very long string of URL that biglyBT tried to query and said timeout when I was blindly trying, but forgot it. 

BTW, does the "Local connections stays in state: connecting" issue related to biglyBT complaining about 60000/tcp refused? I am trying to solve the former problem, leading me down this rabbit hole. If it is not related, then I'll leave it as such and go on. I could still seed with remote connections anyway

**Comment by parg** [comment]
For TCP the NAT test server connects back to the determined public IP address of the client and sends a special message type that the NAT test client is registered for, for UDP it uses a second network interface to send a reply to ensure that the port really is open and it isn't just the case that a router is opening a temporary hole.

The process is identical for IPv4 and IPv6.

If a local connection is stuck "connecting" then it would seem to mean that outbound connections are being prevented by something. Given they stay in that state and don't fail quickly with "connection refused" (for example) then I guess they are timing out (you could turn on logging (Options->Logging) and look at the console (Tools->Console) to see)

**Comment by SodaWithoutSparkles** [comment]
Ah yes they are timing out. Adding "timed out" to console filter revealed that there are tonnes of "request timed out" messages like these:

```
[19:22:16.766] {net} incoming crypto handshake failure: Protocol decode aborted: timed out after 30sec: 24 bytes read, crypto: state=12,sub=1,in=104,out=107
[19:22:16.767] {nwman} Connection [/<IP>:14423] forcibly timed out after 30sec due to socket inactivity
```

There are also warnings about these:
```
[19:26:38.406] {nwman} Incoming stream from [/<IP>:32491] does not match any known byte pattern: 13426974 546F7272 656E7420 70726F74 6F636F6C 00000000 00180005 6DF99344 A488592B A3F71FFA EF964A1D BFE0C8ED 2D714234 3435302D 43694E48 6E583372 21595255 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000...
```

Maybe these two are related, and the TCP failed is another problem? I really hates debugging these issues where multiple symptoms mix together.

**Comment by parg** [comment]
"Not matching a byte pattern" is generally a symptom of you removing/stopping a download and peers still trying to connect to you due to the de-registration process not being perfect and there being stale peer data "out there"

**Comment by SodaWithoutSparkles** [comment]
That make sense. I did have a torrent that was stopped. But the first 2 were abnormal? I also found several possibily related log entries by pausing the logs and enabling real-time update: 

```
[20:16:09.716] {stdout} DEBUG::<Date>  Conflicting peer identification: unknown_client [LTEP]: "Unknown [MG-3.01.4275KYgPk4dl]" / "MediaGet2 3.01.4275" [4D656469614765743220332E30312E34323735], Peer ID: 4D472D332E30312E343237354B5967506B34646C
[20:16:06.432] {plug} [Distributed DB] NATPuncher: Received tunnel inbound message from 42D99F85...[/<IP>:65266,V54]
[20:16:09.258] {stdout} DEBUG::<Date>  Unable to decode peer correctly - peer ID bytes: 2D4B4732323132332D00424D466B4D666D575A68
[20:15:49.767] {plug} [Distributed DB] requestNodeAdd for 72B801BB...[hba=Y,bad=0,OK=20657040] denied as too soon after previous 
[20:15:33.481] {net} Idle limit exceeded for loc=60000 - /<IP>:53526, removing
```

**Comment by SodaWithoutSparkles** [comment]
Would a debg.zip help? If so, how can I send you that file? Ofc I can do torrent, or maybe password protect it, upload here, then email you a PW?

**Comment by parg** [comment]
You could try sending to paul@biglybt.com - might have to rename it to (to, say, debugzip)  get through mail scanners

**Comment by SodaWithoutSparkles** [comment]
> You could try sending to paul@biglybt.com - might have to rename it to (to, say, debugzip)  get through mail scanners

Sent. Titled `Github issue #2933 debug`

**Comment by parg** [comment]
So you appear to have a VPN active (10.217.*.*) and yet have BiglyBT bound to your LAN IP (192.168.1.*)  ?

**Comment by parg** [comment]
And you're proxying peer data connections through a SOCKS server?

Also, unset "IPV6 Ignore Network Addresses", it doesn't do anything very useful

**Comment by SodaWithoutSparkles** [comment]
Disabled the "ignore ipv6" option
I did not have a socks proxy, that might be an mistake when I am playing around with the controls to try and resolve the issue.

**Comment by parg** [comment]
Disable Options->Connection->Proxy: Enable proxying of peer communications (outgoing connections only) [restart required]

**Comment by SodaWithoutSparkles** [comment]
Edited my response.

Once in a while the log said "failed to establish outgoing connection: transport closed". This one is probably due to the connection limit. Maybe add a state like `max concurrent connections` or `saturated outlinks`.

**Comment by parg** [comment]
So outgoing connections are now working since you disabled your non-existant SOCKS proxy? That's why I don't like to get involved in these kind of issues.

**Comment by SodaWithoutSparkles** [comment]
Thanks for your help. Although the "port forwarding seems not working" issue is not solved, the original problem prompting me to jump into the rabbit hole is. I will wait for a few hours until all dusts are settled, but it would probably be fine.

**Comment by SodaWithoutSparkles** [comment]
> So outgoing connections are now working since you disabled your non-existant SOCKS proxy? That's why I don't like to get involved in these kind of issues.

Not all, but most

should the UI warn the user about a non-existant SOCKS proxy? Like a check that needs to be done before saving. 

For example, when adding a hidden wifi network, you cannot leave the password field empty if you selected an encryption mode. When manually setting a proxy in an android phone (wifi > proxy), if you dont enter a domain and port, it would not let you save. 

This sounds like a thing that it should have a check for. Nothing too crazy, just adding a format check for "at least 1 `.`" and a field presence check for port.

**Comment by parg** [comment]
Or, if you go into advanced settings, don't randomly set stuff and expect things to work.

**Comment by SodaWithoutSparkles** [comment]
Thank you for helping me. And I apologize for wasting your time.

**Comment by parg** [comment]
nah, no problem :) glad things improved a bit for you!

**Comment by parg** [comment]
I added " (SOCKS)" to the outbound peer connection state to make it a bit more obvious what's going on

**Comment by SodaWithoutSparkles** [comment]
I might have stumbled upon the real reason of "many local connections stay in state connecting". Today I tried to look at the logs, now that I know how to filter properly (filter out any logs containing `piece`). And I found out there is a "network address unreachable" error. 

I then tried to ping the IP address using my phone, the rpi that biglyBT was on, and a separate online ping checker ([link](https://check-host.net/)). None of them returned any response, which means the host is probably down, or behind double-NAT, or their ISP filtered it, or they misconfigured something. Either way, its none of my business.
